### ISSUE TITLE
create integration tests

### ISSUE THREAD

**Issue by hohwille** [oasp/oasp4j#385]
We already did some invest for tests on higher levels of integration. However, we had some trouble to make them run reliable in CI.
Now that we have switched to spring-boot (see #232) we can easily recover our previous work and do some improvements. I already prepared some things for that.
See also:
https://docs.spring.io/spring-boot/docs/current/reference/html/howto-embedded-servlet-containers.html#howto-user-a-random-unassigned-http-port

We can then annotate our tests like this:

```
@SpringApplicationConfiguration(classes = MyApplication.class)
@WebIntegrationTest
public class TablemanagementRestServiceTest extends SubsystemTest {

   @Value("${local.server.port}")
  private int port;

  ...
}
```

This will start our app in a tomcat on an available random port and allows us to run REST-Requests towards the server as a full integration test of the server backend.

**Comment by hohwille** [comment]
Within such test, we can create a REST-client from our service interface (that we also need to extract from our ```*RestServiceImpl``` classes (and move to ```api``` module - see #384).
Then we can create a test-infrastructure (in ```oasp4j-module-test```) with code like this:
```
    JAXRSClientFactoryBean factoryBean = new JAXRSClientFactoryBean();
    factoryBean.setAddress("http://localhost:" + this.port + "/restaurant/services/rest");
    factoryBean.setHeaders(new HashMap<String, String>());
    // example for basic auth
    String payload = user + ":" + password;
    String authorizationHeader = "Basic " + Base64Utility.encode(payload.getBytes());
    factoryBean.getHeaders().put("Authorization", Arrays.asList(authorizationHeader));
    factoryBean.setProviders(Arrays.asList(this.jacksonJsonProvider));
    return factoryBean;
```

With that we can do
```
    JAXRSClientFactoryBean factoryBean = createJaxRsClientFactory(user, password);
    factoryBean.setServiceClass(TablemanagementRestService.class);
    return factoryBean.create(TablemanagementRestService.class);
```

then we have an instance of our REST-Service interface that we can use to call REST-Services on our server running locally in the JUnit test. The Java objects will be serialized to JSON and HTTP request is triggered, result is retrieved and de-serialized from JSON and provided back as Java object.

**Comment by hohwille** [comment]
Further we can integrate with ```com.jayway.jsonassert.JsonAssert``` and get the raw payload from CXF client to do assertions on JSON structure. This allows us to write tests that ensure that REST-API on JSON level stays compatible for JS client (what easily breaks).

**Comment by hohwille** [comment]
There is a problem with my suggestion:
* osap4j does not use basic authentication but requires form base auth what is generally sick for REST-Services.
* Form based auth is IMHO not supported by CXF but we can do research on this. However using cookies etc. is not really the way to go for REST communication between backends.

So...
* We should switch over to basic auth or at least allow to use it in test (maybe via spring profile)
* Should we use or support basic auth generally in our sample?

**Comment by hohwille** [comment]
FYI:
If you locally want to first want to hack it so you get the test working you can start here:
https://github.com/oasp/oasp4j/blob/develop/samples/core/src/main/java/io/oasp/gastronomy/restaurant/general/configuration/WebSecurityConfig.java#L123

But before creating a pull-request (https://github.com/oasp/oasp4j/wiki/oasp-code-contributions) we need a real fix for this...

**Comment by maihacke** [comment]
Maybe we could configure an additional filter which uses an http-basic header if present?

**Comment by hohwille** [comment]
> Maybe we could configure an additional filter which uses an http-basic header if present?

That is exactly what was done in the code. Now we have this only active for JUnits to not accidently interfere the security (see #406).

**Comment by hohwille** [comment]
We already made some progress here. However, my further thoughts for improvements with the next steps:
* extract an abstract base class that REST service integration tests derive from. This should also include the port number injection, etc.
* move test infrastructure to one or multiple spring beans configured via a @Configuration class that has the junit test profile. This allows better reuse. We might need further infrastructure for testing and that might be reused without being forced to extend a predefined JUnit test base class already bound to a specific test stage (category). That test infrastructure can be injected into the base test class and thereby made available to subclasses. Currently we use a static method for that what is too unflexible.
* We should revisit the name `OaspProfile`. I would rather name it `SpringProfileConstants`.
* I would have constants this way:
```
public static final String JUNIT = "junit";
public static final String NOT_JUNIT = "!" + JUNIT;
```

Instead of adding the Junit profile via @Profile annotation to a Test-Case I would add it to `src/test/resources/config/application.properties` like this
```
spring.profiles.active=junit,dev,...
```

**Comment by hohwille** [comment]
`@WebIntegrationTest("server.port=0")`
for simplicity I would recommend to use
`@WebIntegrationTest` in the test and
```
server.port=0
```
to `src/test/resources/config/application.properties`
This is a general strategy to use the random port infrastructure. If a test-case shall start the app on a HTTP port this should be a free random port as everything else will mess up automated builds in CI (jenkins, travis, etc.). Still if some some strange and exotic reason someone would want to use a different port we could still override via annotation in his test, but I would consider this as a pure anti-pattern.

**Comment by jomora** [comment]
> - We should revisit the name `OaspProfile`. I would rather name it `SpringProfileConstants`.
> - I would have constants this way:
> 
> ```
> public static final String JUNIT = "junit";
> public static final String NOT_JUNIT = "!" + JUNIT;
> ```
> 
> Instead of adding the Junit profile via @Profile annotation to a Test-Case I would add it to `src/test/resources/config/application.properties` like this
> 
> ```
> spring.profiles.active=junit,dev,...
> ```

@hohwille I found a funny problem regarding the `NOT_JUNIT` constant.
Using an exclamation mark ("!") at the beginning of property values (like "!junit") lets Spring treat them as comments, i.e. they are not taken into account during an app's startup.

Consider the following example: 
Currently our `WebSecurityConfig` looks like follows: 
```java
@Configuration
@EnableWebSecurity
@Profile(SpringProfileConstants.NOT_JUNIT)
public class WebSecurityConfig extends BaseWebSecurityConfig { ... }
```
The following constant is used: 
```java
public static final String NOT_JUNIT = "!" + JUNIT;
```
By setting a breakpoint in the default constructor and using the debugger, I found out that `WebSecurityConfig` is not initialized during startup.
Also, setting `spring.profiles.active=h2mem,!junit` in `application.properties` results in profile `!junit`  being _inactive_.
However, by using a different name ,e.g. `not_junit`, everything worked as expected regarding both `@Profile` annotation as well as `application.properties`. 

Therefore, we should review the naming of constants in `SpringProfileConstants.java`.

**Comment by jomora** [comment]
@hohwille I think everything here has been taken into account. We should close this issue.

**Comment by kiran-vadla** [comment]
@hohwille , Kindly look into the ask from @jomora when you get a minute .

**Comment by kiran-vadla** [comment]
This will be addressed in the Spring Boot 1.5 Upgrade (oasp4j-3.0.0)

**Comment by AbhayChandel** [comment]
Points in this issue have been covered and hence can be closed.

**Comment by kiran-vadla** [comment]
Closing this issue as the changes requested as part of this issue are addressed in PR#583.

Thanks,
@kiran-vadla
